File: 0403-Frog Jump/0403-Frog Jump.cpp
class Solution {
public:
    bool canCross(vector<int>& stones) {
        unordered_map<int, unordered_set<int>> steps;
        steps[0].insert(0);
        for (int stone : stones)
        {
            for (int step : steps[stone])
            {
                if (step > 1)
                {
                    steps[stone + step - 1].insert(step - 1);
                }
                steps[stone + step].insert(step);
                steps[stone + step + 1].insert(step + 1);
            }
        }
        
        return steps[stones.back()].size() > 0;
    }
};

File: 0401-0500/0403-Frog Jump/0403-Frog Jump.py
import collections
class Solution:
    def canCross(self, stones: List[int]) -> bool:
        steps = collections.defaultdict(set)
        steps[0].add(0)
        for stone in stones:
            for step in steps[stone]:
                if step > 1:
                    steps[stone + step - 1].add(step - 1)
                steps[stone + step].add(step)
                steps[stone + step + 1].add(step + 1)
        return len(steps[stones[-1]]) > 0
